crates/red_knot_python_semantic/src/types.rs

```diff
@@ -6102,6 +6095,85 @@ impl<'db> FunctionType<'db> {
     }
 }
 
+// This is a separate block for now to avoid making `to_overloaded` a salsa query. This is a
+// temporary workaround until salsa supports returning `Option<&T>` from a tracked query.
```

Comment:
Is this needed because salsa woudd otherwise incorrectly pick up the inner function and make it a method?

Reply:
I'm not sure by "make it a method" but I think it would make it a query because the previous `impl` block is marked with `#[salsa::tracked]` and it seems wasteful to only cache the `.as_ref` call because the inner function is already tracked.

I could possibly avoid marking the inner function as salsa tracked and merge this into the above `impl` block.

Reply:
> I'm not sure by "make it a method" but I think it would make it a query because the previous impl block is marked with #[salsa::tracked] and it seems wasteful to only cache the .as_ref call because the inner function is already tracked.

The `#[salsa::tracked]` on the impl block doesn't make all functions in that block query. Only functions that itself are also marked with `#[salsa::tracked]` will become queries. That's why I think it should be fine to merge the impl blocks.

Reply:
Yeah, I'm not sure why I thought of that. I just realized after sending this message. Thanks. I'll change it.